storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java

```diff
@@ -16,44 +16,2260 @@
+                            : Optional.empty();
+
+                        return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, curLocalLogStartOffset, epochResult));
+                    } else if (targetTimestamp == ListOffsetsRequest.LATEST_TIMESTAMP) {
+                        return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, logEndOffset(), leaderEpochCache.latestEpoch()));
+                    } else if (targetTimestamp == ListOffsetsRequest.LATEST_TIERED_TIMESTAMP) {
+                        if (remoteLogEnabled()) {
+                            long curHighestRemoteOffset = highestOffsetInRemoteStorage();
+                            OptionalInt epochOpt = leaderEpochCache.epochForOffset(curHighestRemoteOffset);
+                            Optional<Integer> epochResult = epochOpt.isPresent()
+                                ? Optional.of(epochOpt.getAsInt())
+                                : curHighestRemoteOffset == -1
+                                    ? Optional.of(RecordBatch.NO_PARTITION_LEADER_EPOCH)
+                                    : Optional.empty();
+                            return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, curHighestRemoteOffset, epochResult));
+                        } else {
+                            return new OffsetResultHolder(new FileRecords.TimestampAndOffset(RecordBatch.NO_TIMESTAMP, -1L, Optional.of(-1)));
+                        }
+                    } else if (targetTimestamp == ListOffsetsRequest.MAX_TIMESTAMP) {
+                        // Cache to avoid race conditions.
+                        LogSegment latestTimestampSegment = null;
+                        for (LogSegment segment : logSegments()) {
+                            if (latestTimestampSegment == null) {
+                                latestTimestampSegment = segment;
+                            } else if (segment.maxTimestampSoFar() > latestTimestampSegment.maxTimestampSoFar()) {
+                                latestTimestampSegment = segment;
+                            }
+                        }
+                        // cache the timestamp and offset
+                        TimestampOffset maxTimestampSoFar = latestTimestampSegment.readMaxTimestampAndOffsetSoFar();
+                        // lookup the position of batch to avoid extra I/O
+                        OffsetPosition position = latestTimestampSegment.offsetIndex().lookup(maxTimestampSoFar.offset);
+                        Optional<FileRecords.TimestampAndOffset> timestampAndOffsetOpt =
+                                StreamSupport.stream(latestTimestampSegment.log().batchesFrom(position.position).spliterator(), false)
+                                    .filter(b -> b.maxTimestamp() == maxTimestampSoFar.timestamp)
+                                    .findFirst()
```

Comment:
TODO: This always iterates over the full collection, whereas the Scala `findFirst()` method returns as soon as an element matching the predicate is found.